ui/admin-portal/src/app/components/util/joblink/joblink.component.ts

```diff
@@ -14,107 +14,118 @@
+  removeJobRequest: boolean;
+
+  //Heading displayed for job search - depends on whether there is already a job associated
+  searchHeading: string;
+
+  //True if searching for jobs
+  searching: boolean;
+
+  constructor(private jobService: JobService) {
   }
 
   ngOnInit(): void {
 
-    this.form = this.fb.group({
-      sfJoblink: [this.joblink,
-        [Validators.pattern(`${salesforceUrlPattern}|${salesforceSandboxUrlPattern}`)], //Sync validators
-        [this.sfJoblinkValidator()] //Async validators
-      ],
-    });
+    //See https://ng-bootstrap.github.io/#/components/typeahead/examples
+    this.doJobSearch = (text$: Observable<string>) =>
+    text$.pipe(
+      debounceTime(300),
+      distinctUntilChanged(),
+      tap(() => {
+        this.searching = true;
+      }),
+      switchMap(text =>
+        this.jobService.searchPaged(
+          {keyword: text, jobNameAndIdOnly: true, pageSize: 10}).pipe(
+          map(result => result.content),
+          catchError(() => {
+            return of([]);
+          }))
+      ),
+      tap(() => this.searching = false)
+    );
   }
 
-  private sfJoblinkValidator(): AsyncValidatorFn {
-    return (control: AbstractControl): Observable<ValidationErrors | null> => {
-      const url: string = control.value;
-      let retval;
-
-      this.updateError.emit(null);
-
-      if (url == null || url.length === 0) {
-        //Empty url always validates
-        retval = of(null)
-        this.joblinkValidation.emit("");
-      } else {
-        //See if we have name for a job corresponding to this url
-        retval = this.salesforceService.getOpportunity(url).pipe(
-          //As side effect populate the job details
-          tap(opportunity => {
-            const valid = opportunity && opportunity.name !== null;
-            const validationEvent = new JoblinkValidationEvent(valid);
-            if (valid) {
-              validationEvent.sfJoblink = this.form.controls.sfJoblink.value;
-              validationEvent.jobname = opportunity.name;
-            }
-            this.joblinkValidation.emit(validationEvent);
-          }),
-
-          //Null names turn into validation error - otherwise no error
-          map(opportunity => opportunity === null ? {'invalidSfJoblink': true} : null),
-
-          //Problems connecting to server will be displayed but we won't
-          //treat it as a validation error
-          catchError(err => {
-            this.updateError.emit(err);
-            return of(null);
-          })
-        );
-      }
+  ngOnChanges(changes: SimpleChanges): void {
+    //Display appropriate heading for job search
+    this.searchHeading = (this.jobId ? "Change" : "Optional") + " job association";
 
-      return retval;
-    };
+    //If there is already a job associated, get its name and construct the default
+    //job request (ie to retain existing job)
+    if (this.jobId) {
+      this.jobService.get(this.jobId).subscribe({
```

Comment:
Good to unsubscribe this when the component is destroyed.

```
private jobSubscription: Subscription;

ngOnChanges(changes: SimpleChanges): void {

  . . .

  // ngOnChanges may be called many times, so unsubscribe from the previous subscription if it exists?
  if (this.jobSubscription) {
    this.jobSubscription.unsubscribe();
  }

  this.jobSubscription = this.jobService.get(this.jobId).subscribe({
    // subscription logic
  });
}


ngOnDestroy(): void {
  // Ensure all subscriptions are unsubscribed when the component is destroyed
  if (this.jobSubscription) {
    this.jobSubscription.unsubscribe();
  }

```

Reply:
Hi @sadatmalik I know it is a little surprising, but the job service subscription is to an Http Observable - and you don't need to unsubscribe from them. See https://angular.io/guide/http-request-data-from-server#:~:text=In%20general%2C%20you%20should%20unsubscribe,request%20responds%20or%20times%20out.

Reply:
Thanks for clarifying that @camerojo - approved for merging.